Added Seurat <-> SCE conversion sections (#668)
AlexsLemonade/training-modules: module-cheatsheets/scRNA-seq-advanced-cheatsheet.md

## `Seurat` and `SCE` object conversion

The `Seurat` documentation provides a [vignette about converting objects](https://satijalab.org/seurat/articles/conversion_vignette.html) between `SCE` and `Seurat` formats.

In addition, we provide some code examples below for how you can accomplish these conversions.
For all code examples below, it is assumed that the `SingleCellExperiment` library has been loaded into your R environment:

```r
library(SingleCellExperiment)
```

### Converting from `Seurat` to `SCE`

The following example code assumes you have a `Seurat` object called `seurat_obj`.

```r
# Convert Seurat object to SCE object
sce_object <- Seurat::as.SingleCellExperiment(seurat_obj)
```

By default, all assays present in the `Seurat` object will be ported into the new `SCE` object.

> To only specify that certain assays are retained, you can optionally provide the argument `assays`, as in: `assays = c("assays", "to", "keep")`.

Author: The caveat is now a big table and bullet list of sce/seurat bookkeeping. There's a good pun/joke to be made with "confusion matrix" here.

### Converting from `SCE` to `Seurat`

The following example code assumes you are starting with an `SCE` object called `sce_object`.

The function `Seurat::as.Seurat()` can be used to convert an `SCE` object into a `Seurat` object and takes the following arguments:

- The `SCE` object to convert
- Optional named arguments with the following defaults:
  - `counts = "counts"` specifies that the `SCE` object contains a `"counts"` assay of normalized counts that should be included during conversion.
    - If there is no `"counts"` assay in the SCE object, set this argument as `counts = NULL` or rename accordingly, e.g. `counts = "whatever_assay_name_you_are_using"`.
  - `data = "logcounts"` specifies that the `SCE` object contains a `"logcounts"` assay of normalized counts that should be included during conversion.
    - If there is no `"logcounts"` assay in the SCE object, set this argument as `data = NULL` or rename accordingly, e.g. `data = "whatever_assay_name_you_are_using"`.
  - `assay = NULL` specifies that, by default, all assays will be converted.

>     - To specify that an additional assay besides `"counts"` or `"logcounts"` should be converted, include it here as in `assay = "additional_assay_name"`.

Reviewer: I think I'd flip the order of these two instructions. Because what you are really saying is that if you want to _exclude_ some assays, you need to specify this argument, with all of the assays you want to include (which is confusing). I think it may be less confusing to say something like "The `assays` argument allows you to specify specific addtional assays to include. By default, all assays are converted..." Probably also worth saying if you _don't want additional assays you need to set this to `c()` which is non-intuitive.

Author: `assay` apparently does not always mean what we think it means, so I changed this in a different way!

# Convert SCE object to Seurat object, where the SCE object
#  contains a `counts` but not a `logcounts` assay
seurat_object <- Seurat::as.Seurat(sce_object, data = NULL)

```

#### Approaches from `ScPCA`

In addition, this [documentation from the `ScPCA`](https://scpca.readthedocs.io/en/latest/faq.html#what-if-i-want-to-use-seurat-instead-of-bioconductor) introduces how to convert `SCE` objects to `Seurat` objects.
Although this documentation was written for `ScPCA` datasets, the steps generally apply to any `SCE` object.

We also offer a conversion function `sce_to_seurat()` as part of our [`scpcaTools()` package](https://github.com/AlexsLemonade/scpcaTools/), which holds utilities used in the `ScPCA` workflow.
Again, although this function was written to convert `SCE` objects from `ScPCA`, it should generally work for most `SCE` objects, although it will only retain a single assay (raw `"counts"`) in the new `SCE` object, and it will not retain reduced dimension representations (e.g., PCA or UMAP).
Therefore, this function is mostly useful at the early stages of processing before you have normalized counts and and calculated reduced dimensions.

> You can obtain this package using the `remotes` package, which may also need to be installed first:

Reviewer: I don't think we should recommend this at this stage, for a bad reason and a good reason. The bad reason is that I'm just not confident in it yet, and don't want to be debugging it as a general tool at the moment.

The good reason is that we are using R 4.1 on the server, and `scpcaTools` now requires R 4.2, so this won't actually work!

I think it is fine to point people to the code as an example though!

Author: Purged `scpcaTools` for R-versioning reasons, and reorganized the remaining text.

The table below shows different aspects of single-cell objects and how to access the associated data, assuming the default names for each type of single-cell object.
Importantly, these default names may not be the same - for example, `SCE` refers to normalized counts as `"logcounts"`, but `Seurat` refers to normalized counts as `"data"`.
The table below additionally assumes that the `Seurat` object assay of interest is named `"RNA"`, which is the default assay name that `Seurat` assigns.

| Data aspect | `SCE` | `Seurat` |
|------------|---------|---------|
| Raw counts assay | `counts(sce_object)` | `seurat_obj[["RNA"]]@counts` |
| Normalized counts assay | `logcounts(sce_object)` | `seurat_obj[["RNA"]]@data` |
| Reduced dimension: PCA matrix | `reducedDim(sce_object, "PCA)` | `seurat_obj$pca@cell.embeddings` |
| Reduced dimension: UMAP matrix | `reducedDim(sce_object, "UMAP)` | `seurat_obj$umap@cell.embeddings` |
| Cell-level metadata | `colData(sce_object)` | `seurat_obj@meta.data` |
| Feature (gene)-level metadata | `rowData(sce_object)` | `seurat_obj[["RNA"]]@meta.features`|

> | Miscellaneous additional metadata | `metadata(sce_object)` | `seurat_obj@misc`|

Reviewer: 👏🏼 👏🏼 👏🏼 

This I think will be really useful! (My only fear is that it will make people ask more for the "Seurat equivalent" for all the things.)

Author: If anything it's useful for me!

### `Seurat` and `SCE` object conversion

> This [vignette](https://satijalab.org/seurat/articles/conversion_vignette.html) from the `Seurat` documentation explains how to convert between `SCE` and `Seurat` formats.

Reviewer: Rewriting for a longer, more informative link. 

```suggestion
The `Seurat` documentation provides a [vignette about converting objects](https://satijalab.org/seurat/articles/conversion_vignette.html) between `SCE` and `Seurat` formats.
```

I think it is right to link to this, but boy is it sparse. And boy do I not love instructions to install a package from the `develop` branch.

Author: > And boy do I not love instructions to install a package from the `develop` branch.

I bet you'll love even more how there is no `develop` branch anymore in that repo, and there are no instructions for installing the other `Seurat` package dependency which lives in a _different_ account from that one IIRC.

So, yeah, this "vignette" has problems, and I'm not at all opposed to removing it.

In addition, we provide some code examples below for how you can accomplish these conversions.
For all code examples below, it is assumed that `SingleCellExperiment` and `Seurat` libraries have been loaded in your R environment:

```r
library(Seurat)
library(SingleCellExperiment)

> ```

Reviewer: I think we may not want to do this, or we may want to only half do this. I think it is important to know which of the functions we are using is from `Seurat` vs. SCE, so I might load SCE, but leave `Seurat` functions prefixed with `Seurat::`

Author: Done in https://github.com/AlexsLemonade/training-modules/pull/668/commits/8ce618227f1993746208c2cd57611dc20f700112

#### Converting from `Seurat` to `SCE`

The following example code assumes you have a `Seurat` object called `seurat_obj`.

```r
# Convert Seurat object to SCE object using the
#  Seurat function `as.SingleCellExperiment()
sce_object <- as.SingleCellExperiment(seurat_obj)
```

> Alternatively, you can extract individual slots from the `Seurat` object and build your `SCE` object from scratch.

Reviewer: I feel like introducing this without a reason _why_ we want to do it is not particularly helpful. Is there any case where we would want to extract just one assay with no metadata? If there is a specific case that we feel is worth illustrating, we should do that, but I don't want to encourage people to drop metadata as part of the conversion. 

If there is something that the conversion does not handle automatically, we might want to show how to add that to an existing SCE, or rename things that might end up in unexpected locations, but I'm not sure linking to building an SCE is useful here, especially without the matching docs for how to extract each part of a Seurat object.

Author: Hm, maybe this isn't needed at all. I included it as a way to offer some kind of direction if conversion goes wrong, but really if conversion goes _wrong_, a few small details won't help (that's what consultations/office hours are more reasonably for).
Based on the [Seurat source](https://github.com/satijalab/seurat/blob/763259d05991d40721dee99c9919ec6d4491d15e/R/objects.R), it seems like all the bits do get converted, so I'm thinking just call it a day with `Seurat::as.SingleCellExperiment()`.

You can refer to the ["The `SingleCellExperiment` class" chapter in _Orchestrating Single Cell Analyses_](http://bioconductor.org/books/release/OSCA.intro/the-singlecellexperiment-class.html) for approaches to constructing `SCE` objects.
For example, to obtain an SCE object that contains only the raw counts stored in the `Seurat` object's "RNA" assay:

```r
# First, extract the counts matrix from the Seurat object
counts_matrix <- GetAssayData(seurat_obj[["RNA"]])

# Create an SCE object from the counts matrix
sce_object <- SingleCellExperiment(assays = list(counts = counts_matrix))
```

#### Converting from `SCE` to `Seurat`

This [documentation from the `ScPCA`](https://scpca.readthedocs.io/en/latest/faq.html#what-if-i-want-to-use-seurat-instead-of-bioconductor) introduces how to convert `SCE` objects to `Seurat` objects.
Although this documentation was written for `ScPCA` datasets, the steps generally apply to any `SCE object`.

> Briefly, here is how you can convert a `Seurat` to `SCE` object, focusing on porting over _assays_.

Reviewer: I think we will to show the steps to add at least the cell and feature metadata here. We present it as "optional" in the docs, but we probably shouldn't.

Author: Something else I also thought about is whether we should straight-up recommend `scpcaTools::sce_to_seurat()` here, in addition to or instead of individual steps.

Author: Hmm, maybe not  ^, since our function only allows _one assay_ to get ported over... But I've stubbed something out already so we'll see.

Reviewer: I think you can point people to it as an example, but I would not exactly _recommend_ it.

We provide `"RNA"` as the argument to `assay`, as this `Seurat`'s default name for raw count matrices.

> ```r

Reviewer: Spacing
```suggestion

```r
```

# Create seurat object from the existing `sce_object`'s counts matrix,
#  using the Seurat function `CreateSeuratObject`
seurat_object <- CreateSeuratObject(counts = counts(sce_object),
                                    assay = "RNA",
                                    project = "name of your project goes here")
```

> 

Reviewer: We may want to note that we are not using `Seurat::as.Seurat()` though that may be an option (it didn't used to work very well, but maybe works better now?)

Author: I'll check it out on whatever versions we have in the training `renv`, haven't use that function yet.
